subsys/bluetooth/host/att.c

```diff
@@ -2948,6 +2987,55 @@ static int bt_att_accept(struct bt_conn *conn, struct bt_l2cap_chan **ch)
 BT_L2CAP_CHANNEL_DEFINE(att_fixed_chan, BT_L2CAP_CID_ATT, bt_att_accept, NULL);
 
 #if defined(CONFIG_BT_EATT)
+static k_timeout_t credit_based_connection_delay(struct bt_conn *conn)
+{
+	/*
+	 * 5.3 Vol 3, Part G, Section 5.4 L2CAP COLLISION MITIGATION
+	 * ... In this situation, the Central may retry
+	 * immediately but the Peripheral shall wait a minimum of 100 ms before retrying;
+	 * on LE connections, the Peripheral shall wait at least 2 *
+	 * (connPeripheralLatency + 1) * connInterval if that is longer.
+	 */
+
+	if (conn->role == BT_CONN_ROLE_CENTRAL) {
+		return K_NO_WAIT;
+	}
+
+	uint32_t rand_delay = sys_rand32_get() & 0x7; /* Small random delay for IOP */
+	uint32_t calculated_delay = 2 * (conn->le.latency + 1) * conn->le.interval;
+
+	return K_MSEC(MAX(100, calculated_delay + rand_delay));
+}
+
+int att_schedule_eatt_connect(struct bt_conn *conn, struct att_connection_work_context context)
+{
+	struct bt_att *att = att_get(conn);
+
+	att->connection_work_context = context;
+	return k_work_schedule(&att->connection_work, credit_based_connection_delay(conn));
+}
+
+void ecred_connect_cb(struct bt_conn *conn, uint16_t result, uint8_t attempted_to_connect,
+		      uint8_t succeded_to_connect)
+{
+	struct bt_att *att = att_get(conn);
+	struct att_connection_work_context context;
+
+	if (result == att->previous_ecred_connection_result &&
```

Comment:
Both the peer has to respond with "no resources" and we must respond with "no resources" to a request from the peer for it to be considered a collision. This checks that "no resources" is the status twice in a row.

Changing it a bit to make it more explicit instead of comparing the previous and current results.